bin/faucet/src/faucet/mod.rs

```diff
@@ -0,0 +1,422 @@
+                    && err.message().contains("incorrect initial state commitment") =>
+            {
+                self.handle_desync(err.message()).with_context(|| {
+                    format!("failed to recover from desync error: {}", err.message())
+                })
+            },
+            // TODO: Look into which other errors should be recoverable.
+            //       e.g. Connection error being lost to RPC client is probably not fatal.
+            others => Err(others).context("failed to handle error"),
+        }
+    }
+
+    /// Attempts to rollback back local state to match that indicated by the node.
+    ///
+    /// This relies on parsing the stringified error
+    /// `VerifyTxError::IncorrectInitialAccountCommitment`.
+    ///
+    /// Returns an error if the rollback was unsuccesful. This should be treated as fatal.
+    fn handle_desync(&mut self, err: &str) -> anyhow::Result<()> {
+        let onchain_state = parse_desync_error(err).context("failed to parse desync message")?;
+
+        // Find the matching local state, unless we've dropped it already.
+        let rollback = self
+            .prior_state
+            .iter()
+            .position(|state| state.commitment() == onchain_state)
+            .context("no matching local state to rollback to")?;
+
+        // Rollback the local state.
+        // SAFETY: The index must exist since the element was just found.
+        self.data_store.update_faucet_state(self.prior_state[rollback].clone());
+        self.prior_state.drain(rollback..);
+
+        tracing::warn!(rollback.count = rollback, "desync detected and handled");
+
+        Ok(())
+    }
+
+    /// Fully handles a single request _without_ changing local state.
+    ///
+    /// Caller should update the local state based on the returned result.
+    async fn handle_request(
+        &self,
+        request: MintRequest,
+        rng: impl FeltRng,
+        rpc_client: &mut RpcClient,
+    ) -> MintResult<(AccountDelta, BlockNumber, Note)> {
+        // Generate the payment note and compile it into our transaction arguments.
+        let p2id_note = P2IdNote::build(self.faucet_id(), &request, rng)?;
+        let tx_args = p2id_note.compile()?;
+
+        let tx = self.execute_transaction(tx_args)?;
+        let account_delta = tx.account_delta().clone();
+
+        let tx = Self::prove_transaction(tx)?;
+        let block_height = self.submit_transaction(tx, rpc_client).await?;
+
+        Ok((account_delta, block_height, p2id_note.into_inner()))
+    }
+
+    fn execute_transaction(&self, tx_args: TransactionArgs) -> MintResult<ExecutedTransaction> {
+        // TODO: Is this cheap? Do we need to carry this around with us, or can we just construct
+        //       when needed?
+        TransactionExecutor::new(self.data_store.clone(), Some(self.authenticator.clone()))
+            .execute_transaction(self.id.inner(), BlockNumber::GENESIS, &[], tx_args)
+            .map_err(MintError::Execution)
```

Comment:
This thing isn't `Send` nor `Sync`. Which made using it in async really impossible. For now I just reconstruct it, but is there some better way to actually use it?

Is construction expensive, or is this fine? The data store is also.. really annoying and requires `Arc` for reasons unknown to me?

Reply:
Yea thats a shame. I'm hoping its purely due to the trait objects in `TransactionExecutor` that don't specify `Send` and `Sync`. Unsure if this stack depends on generated gRPC code which seems to be an issue [here](https://github.com/0xMiden/miden-client/pull/868#issuecomment-2848811369).

Reply:
Yeah, adding bounds for `Send` and `Sync` in the traits or trait objects should do the trick here. We did this a couple times on branches we didn't end up merging for other reasons but [should be a simple fix](https://github.com/0xMiden/miden-base/compare/main...tomyrd-send-sync-traits#diff-cdc577a9fd28ef742a84be42bb71bbe8193f7a3d6b0d9f4773affba863334159) if we want to add that separately.

Reply:
That would be great; they're really rather awkward :D 

Along with the `Arc` requirement in there for the datastore.

Reply:
I'm not sure the `Arc` requirement is too easy to get rid of without extra changes. Or did you mean making it a generic instead of a trait object? I have mentioned before that maybe we could make the executor just take a reference to the datastore on `execute_transaction()` instead of owning it. Initially the datastore was used to load the account code before actually executing, but as this is not needed anymore maybe this can change. Or maybe we could do something like [alternative 2 from this comment.](https://github.com/0xMiden/miden-base/issues/938#issue-2614938080)

Reply:
One thing that is not clear to me is why `handle_request()` function needs to be `async`. All requests to `TransactionExecutor::execute_transaction()` should be serialized, and so, by extension, all requests to `handle_request()` should be serialized. Wouldn't this imply that `handle_request()` should be a synchronous function?

Reply:
`handle_request()` calls `async fn submit_transaction()` which I presume involves network I/O. Wouldn't want to block on that.

Reply:
Yeah its only the node IO that is async.

We could move that out to a separate async thread/task again via a channel which would change that. And that might be a good thing since a lot of the work is blocking.

I'll try that out.

I don't really understand why any of these structs need to be stateful so I can't really way in on it.

Reply:
@bobbinth would you have any reservations about doing something like what @tomyrd did on the [PR I linked above](https://github.com/0xMiden/miden-base/compare/main...tomyrd-send-sync-traits#diff-cdc577a9fd28ef742a84be42bb71bbe8193f7a3d6b0d9f4773affba863334159)? Otherwise, I believe we could make this change.